### DISCUSSION TITLE
Dalli and Memcache compatibilty matrix

### DISCUSSION THREAD

**Discussion by richardmcmillen** [petergoldstein/dalli#951] [Ideas]
Thanks for the project! I have a documentation suggestion:

To help maintainers of projects trying to upgrade Dalli clients and Memcache clusters it would be very useful to have a compatibility matrix.

Especially for those occasionality maintaining others code and without experience of Dalli or Memcache (👋) this would be of great benefit and provide some peace of mind rather than hoping or spending time to learn about each project. For 
maintainers and experienced users this may be incredibly obvious, but not so much to the unfamiliar.

**Comment by petergoldstein** [comment]
@richardmcmillen We could in theory add such a matrix to the documentation, but it wouldn't have much.

Basically the binary protocol in memcached has been frozen since the memcached 1.4 version (released in 2009).  And while the binary protocol is in theory deprecated going forward, it hasn't been removed from any version that's been published.

We added support for the new meta protocol in the Dalli 3.x version, but while that protocol is planned to be the go forward protocol for memcached it's still under development (Dalli development actually uncovered a few issues in memcached that were addressed in 1.6.13).

What would you like in the compatibility matrix given the above?

**Comment by richardmcmillen** [comment]
Thanks a lot for your quick response @petergoldstein 

>What would you like in the compatibility matrix given the above?

The reason I came to the repo is because I was maintaining a ruby project, and I am unfamiliar with developments in the memcache and dalli world (though I saw some chatter about a new protocol).

My question I was trying to answer myself was really, can I `bundle update dalli` safely against the current version of memcache I run. As you have pointed out it may depend on the protocol used.

Given the information provided would something like this make sense?

| Dalli (protocols) | Memcache |
|-------------------|----------|
| <3 (binary)       | <=1.6.18   |
| 3.x (binary,meta) | <=1.6.18   |

When new releases come out perhaps it would look like this:

| Dalli (protocols) | Memcache |
|-------------------|----------|
| <3 (binary)       | <=1.6.18   |
| 3.x (binary,meta) | <=1.6.18   |
| >= 4.x (meta)        | >=2.x      |

I only bring this question up as it may be useful to those really unfamiliar and help give them assurance, but only if you think it makes sense wouldn't be a pain to maintain. As you say, there isn't much to it. I don't know how best to represent the information, but I was trying to understand client versions and their compatibility with memcache.

**Reply by petergoldstein** [reply]
So I don't think this adds value when compared to the current documentation.

I found the proposed table confusing, as it seems to indicate that Dalli wouldn't be compatible with memcached 1.6.19 when it comes out.  That's not true.  And memcached patch releases happen on a regular basis (generally every few months).

The reality is that memcached is a slow moving project that still has regular patch releases.  Dalli should work with any memcached released since 2009 - in the last 14 years.  Given that the binary protocol is frozen, it can be expected to work with all versions of memcached until that protocol is removed (which likely won't happen for another 5-10 years, given the widespread use of the binary protocol).

Given the above facts, there's no reason to maintain a compatibility page with memcached at the patch level.  The current documentation of requirements (https://github.com/petergoldstein/dalli/wiki/Requirements-and-Integrations) seems to make the compatibility clear.   From that page:

`Dalli should work with any version of memcached that supports the binary protocol, which as of time of writing includes the 1.4.x, 1.5.x, and 1.6.x versions of memcached.`

It could use an update for the meta protocol, and I'm happy to add that, but I think it's pretty clear on compatibility.  If a 1.7.x or 2.x version of memcached is released that doesn't support the binary protocol, I'll update it there.

**Comment by petergoldstein** [comment]
I've updated the requirements with more clarity around the binary and meta protocols.  I think the current text is better than the proposed compatibility matrix.

**Reply by richardmcmillen** [reply]
Thanks, I agree with all your points and appreciate your updates. Thank you for your time. 🙌
